tao_compiler/mlir/xla/ral/context/common_context_impl_quantization.cc

```diff
@@ -1335,13 +1336,59 @@ MemRefType<int8_t, 2> ral_pdll_qgemm_onednn_s8_s8_s8_f32_per_channel(
   std::vector<int32_t> output_zero_point({resultZeroPoints.data[0]});
 
   ideep::matmul_forward_params param;
+  // TODO: Only calculate the scale & zero_point once.
   ideep::matmul_forward::prepare(
       param, input_t, weight_t, bias_t, output_t, input_scales, weight_scales,
       output_scales, input_zero_point, output_zero_point, 1.0f, 1.0f,
       ideep::attr_t(), ideep::data_type::s8, ideep::lowp_kind::s8s8,
       ideep::engine::cpu_engine());
-  ideep::matmul_forward::compute<true, true>(param, input_t, weight_t, bias_t,
-                                             output_t);
+
+  bool weight_is_const =
+      dictAttr.get("weight_is_const").as<BoolPDLAttr>().getValue();
+  bool bias_is_const =
+      dictAttr.get("bias_is_const").as<BoolPDLAttr>().getValue();
+  if (!isWeightPrePackingEnabled() || !weight_is_const || !bias_is_const) {
+    // TODO: add a template to control whether bias shoule be reordered
+    ideep::matmul_forward::compute<true, true>(param, input_t, weight_t, bias_t,
+                                               output_t);
+  } else {
+    ideep::tensor packed_weight;
+    ideep::tensor packed_bias;
+    std::string unique_name =
+        "disc.ral_pdll_qgemm_onednn_s8_s8_s8_f32_per_channel";
+    auto state = ctx->getOrCreateResource<OnednnGemmState>(
+        unique_name, []() { return new OnednnGemmState; });
+    {
+      std::lock_guard<std::mutex> l(state->mu);
+      auto& packed_weights = state->packed_weight_cache[weight.data];
+      auto& packed_biases = state->packed_bias_cache[bias.data];
+      auto weights_desc = param.pd.weights_desc();
+      auto bias_desc = param.pd.bias_desc();
+      for (auto& tensor : packed_weights) {
```

Comment:
Could these 2 desc lookup `for` loop be encapsulated into `OnednnGemmState`? Similar logic on L1489

Reply:
done, have made the process of getting or creating the packed weight/bias as a member function of the `OnednnGemmState `